flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointStatistics.java

```diff
@@ -0,0 +1,534 @@
+		if (checkpointStats != null) {
+
+			Map<JobVertexID, TaskCheckpointStatistics> checkpointStatisticsPerTask;
+
+			if (includeTaskCheckpointStatistics) {
+				Collection<TaskStateStats> taskStateStats = checkpointStats.getAllTaskStateStats();
+
+				checkpointStatisticsPerTask = new HashMap<>(taskStateStats.size());
+
+				for (TaskStateStats taskStateStat : taskStateStats) {
+					checkpointStatisticsPerTask.put(
+						taskStateStat.getJobVertexId(),
+						new TaskCheckpointStatistics(
+							taskStateStat.getLatestAckTimestamp(),
+							taskStateStat.getStateSize(),
+							taskStateStat.getEndToEndDuration(checkpointStats.getTriggerTimestamp()),
+							taskStateStat.getAlignmentBuffered(),
+							taskStateStat.getNumberOfSubtasks(),
+							taskStateStat.getNumberOfAcknowledgedSubtasks()));
+				}
+			} else {
+				checkpointStatisticsPerTask = null;
+			}
+
+			if (checkpointStats instanceof CompletedCheckpointStats) {
+				final CompletedCheckpointStats completedCheckpointStats = ((CompletedCheckpointStats) checkpointStats);
+
+				return new CheckpointStatistics.CompletedCheckpointStatistics(
+					completedCheckpointStats.getCheckpointId(),
+					completedCheckpointStats.getStatus(),
+					completedCheckpointStats.getProperties().isSavepoint(),
+					completedCheckpointStats.getTriggerTimestamp(),
+					completedCheckpointStats.getLatestAckTimestamp(),
+					completedCheckpointStats.getStateSize(),
+					completedCheckpointStats.getEndToEndDuration(),
+					completedCheckpointStats.getAlignmentBuffered(),
+					completedCheckpointStats.getNumberOfSubtasks(),
+					completedCheckpointStats.getNumberOfAcknowledgedSubtasks(),
+					checkpointStatisticsPerTask,
+					completedCheckpointStats.getExternalPath(),
+					completedCheckpointStats.isDiscarded());
+			} else if (checkpointStats instanceof FailedCheckpointStats) {
+				final FailedCheckpointStats failedCheckpointStats = ((FailedCheckpointStats) checkpointStats);
+
+				return new CheckpointStatistics.FailedCheckpointStatistics(
+					failedCheckpointStats.getCheckpointId(),
+					failedCheckpointStats.getStatus(),
+					failedCheckpointStats.getProperties().isSavepoint(),
+					failedCheckpointStats.getTriggerTimestamp(),
+					failedCheckpointStats.getLatestAckTimestamp(),
+					failedCheckpointStats.getStateSize(),
+					failedCheckpointStats.getEndToEndDuration(),
+					failedCheckpointStats.getAlignmentBuffered(),
+					failedCheckpointStats.getNumberOfSubtasks(),
+					failedCheckpointStats.getNumberOfAcknowledgedSubtasks(),
+					checkpointStatisticsPerTask,
+					failedCheckpointStats.getFailureTimestamp(),
+					failedCheckpointStats.getFailureMessage());
+			} else {
+				throw new IllegalArgumentException("Given checkpoint stats object of type " + checkpointStats.getClass().getName() + " cannot be converted.");
+			}
+		} else {
+			return null;
```

Comment:
pretty sure this would lead to a NullPointerException, since neither the CheckpointStatshandler, not AbstractExceutionGraphhandler, nor the AbstractRestHandler handle the case of the returned value being null.

Reply:
I think it should actually work and output a serialized `null` value. However, I this distinction is not necessary and should be better pulled out of this method.